### PR TITLE
Add quaternion_to_rotation() to theseus.geometry.functional.so3

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: tests/geometry/functional/test_so3.py [hunk 1]
```diff
@@ -84,3 +84,15 @@ def test_compose(batch_size: int, dtype: torch.dtype):
 
     # check analytic backward for the operator
     check_lie_group_function(so3, "compose", TEST_EPS, group0, group1)
+
```
Comment (luisenp) [inline tests/geometry/functional/test_so3.py:87]: Once again duplicated code. You can reuse the main skeleton of all of these by adding some "preprocess" functions and select the right one for each operation you are checking.

**Review by luisenp — APPROVED** [review]
LGTM. Same comment about the tests.

File: tests/geometry/functional/test_so3.py (hunk 1 above)
Comment (fantaosha) [reply tests/geometry/functional/test_so3.py:87]: Create an issue #402.

File: theseus/geometry/functional/so3.py [hunk 2]
```diff
@@ -376,3 +389,125 @@ def backward(cls, ctx, grad_output):
...
+    jac[:, :, :1] = temp[:, 1:].view(-1, 3, 1)
+    jac[:, :, 1:] = hat(temp[:, 1:])
+    jac[:, 0, 1] = -temp[:, 0]
+    jac[:, 1, 2] = -temp[:, 0]
+    jac[:, 2, 3] = -temp[:, 0]
+
+    return [jac], ret
+
+
+class QuaternionToRotation(lie_group.UnaryOperator):
```
Comment (mhmukadam) [inline theseus/geometry/functional/so3.py:479]: Will we add rotation to quaternion later? From our internal doc there are potential users who would like support for quaternions as well as other representations like axis-angle. Not necessary for this PR but we should make sure we are setting up the API in a way that adding those later is easier.

File: theseus/geometry/functional/so3.py [hunk 3]
```diff
@@ -376,3 +389,125 @@ def backward(cls, ctx, grad_output):
 _jcompose_autograd_fn = _jcompose_impl
 
 compose, jcompose = lie_group.BinaryOperatorFactory(_module, "compose")
+
+
+# -----------------------------------------------------------------------------
+# Unit Quaternion to Rotation Matrix
```
Comment (mhmukadam) [inline theseus/geometry/functional/so3.py:395]: Add in comments the convention we are using i.e. w, x, y, z.

**Review by mhmukadam — APPROVED** [review]

**Pull request merged**